### ISSUE TITLE
Downloads not completeing fully

### ISSUE THREAD

**Issue by TheJaymoe** [SKCraft/Launcher#224]
Some of my users have issues when installing for first time AND when updating that some mods not always fully download, so they are "incomplete" and then they get mod rejection when trying to connect.

Forcing update dosnt work either.

Can't find any reason to why this would happen?

**Comment by sk89q** [comment]
Where are the files hosted?

**Comment by TheJaymoe** [comment]
On my website host

**Comment by sk89q** [comment]
What kind of goat is it

**Comment by TheJaymoe** [comment]
Goat? ^^ If you mean which host its https://www.binero.se/

**Comment by sk89q** [comment]
Close enough. That's what my surgeon says 

I don't know if it is the problem yet. How many users have the problem and how many users have it working?

**Comment by TheJaymoe** [comment]
Out of 8 users one had problem since first install and one had troubles after I updated the modpack.

The first user I think might be that he has bad connection, but shouldnt the launcher redownload the "broken" file then instead of just skipping it? 

Or does it only look for the correct name and not size?

**Comment by sk89q** [comment]
Yeah it probably should but I must have not added in despite calculating
checksums and going through the motions.

**Comment by TheJaymoe** [comment]
`protected void installJar(@NonNull Installer installer,
                              @NonNull File jarFile,
                              @NonNull URL url) throws InterruptedException {
        // If the JAR does not exist, install it
        if (!jarFile.exists()) {
            List<File> targets = new ArrayList<File>();

            File tempFile = installer.getDownloader().download(url, "", JAR_SIZE_ESTIMATE, jarFile.getName());
            installer.queue(new FileMover(tempFile, jarFile));
            log.info("Installing " + jarFile.getName() + " from " + url);
        }
    }`

This?

`if (!jarFile.exists())`

**Comment by sk89q** [comment]
Schrodinger's comments here it seems 

In response to your deleted comment the HttpDownloader class should be doing the checksum checking but checksums are currently not passed in when download() is called

**Comment by TheJaymoe** [comment]
So, all I can do is wait for a new release then? :p

**Comment by sk89q** [comment]
Multiple solutions to your problem. Hell could freeze over 

I'll see about getting to it this week. No guarantees though

**Comment by TheJaymoe** [comment]
Haha okey.

But what about the one that has failed on first install? It can be cause of bad connection that it just skips to the next file?

**Comment by sk89q** [comment]
Could be a bad web server that is dropping connections prematurely or probably more likely something wrong between the server and the user (ISP, wifi, etc.)

**Comment by TheJaymoe** [comment]
Another new player just got same on first install, one mod failed to download fully, the wierd thing is that the clients actually starts without complaining, but then when they try to join the server they get mod rejected.

**Comment by sk89q** [comment]
Any common things between the files that are failing?

**Comment by TheJaymoe** [comment]
No nothing, 2 new players joined both got incomplete mods, but different mods

**Comment by TheJaymoe** [comment]
Some new info 3 out of 4 players that had problem have all been running on WiFi.

So I guess that it skips to the next file with a bad connection.

**Comment by sk89q** [comment]
It normally should throw an error and notice that the file was not fully completed but I guess it's just somehow completing
